pkg/backend/httpstate/client/marshal.go

```diff
@@ -17,19 +17,102 @@ package client
 import (
 	"bytes"
 	"encoding/json"
+	"io"
+	"math"
+
+	"github.com/json-iterator/go"
 
 	"github.com/pulumi/pulumi/sdk/v3/go/common/apitype"
 )
 
-// Unlike json.Marshal preserves possible indentation in req.Deployment.
-func marshalVerbatimCheckpointRequest(req apitype.PatchUpdateVerbatimCheckpointRequest) (json.RawMessage, error) {
-	sentinel := []byte(`"*"`)
-	cp := req
-	cp.UntypedDeployment = sentinel
-	pattern, err := json.Marshal(cp)
-	if err != nil {
+var jsonIterConfig = jsoniter.Config{SortMapKeys: true}.Froze()
+
+// Marshals to canonical JSON in the apitype.UntypedDeployment format.
+//
+// Optimized for large checkpoints.
+//
+// Injects newlines to allow efficient textual diffs over the JSON. Textual diffs currently use O(N^2) memory in the
+// number of newlines, so the injection needs to be conservative. Currently it limits to up to 1024 newlines which would
+// result in max 8MB memory use by the algorithm.
+func MarshalUntypedDeployment(deployment *apitype.DeploymentV3) (json.RawMessage, error) {
+	var buf bytes.Buffer
+	md := &marshalUntypedDeployment{deployment}
+	if err := md.Write(&buf); err != nil {
 		return nil, err
 	}
-	f := bytes.ReplaceAll(pattern, sentinel, req.UntypedDeployment)
-	return f, nil
+	return buf.Bytes(), nil
+}
+
+func marshalVerbatimCheckpointRequest(req apitype.PatchUpdateVerbatimCheckpointRequest) (json.RawMessage, error) {
+	// Unlike encoding/json, using jsonIter here will not reindent req.UntypedDeployment, which is what is needed
+	// for the Verbatim protocol.
+	return jsonIterConfig.Marshal(req)
+}
+
+type marshalUntypedDeployment struct {
+	deployment *apitype.DeploymentV3
+}
+
+func (c *marshalUntypedDeployment) Write(w io.Writer) error {
+	cfg := jsonIterConfig
+	stream := cfg.BorrowStream(w)
+	defer cfg.ReturnStream(stream)
+	err := c.writeToStream(stream)
+	return err
+}
+
+func (c *marshalUntypedDeployment) writeToStream(stream *jsoniter.Stream) error {
```

Comment:
I have no comprehension of the incantation order here 😅 For example, what does `WriteMore()` do? Why is it necessary? Maybe a single line comment somewhere at the top of the method would help? Not a big deal either way, I feel like someone coming across this is going to have to read the API docs anyway.

Reply:
WriteMore writes a "," .. this is a pretty low-level API.

Reply:
😆